platform: nrf5340: target-cfg header

ioannisg opened the pull request:
Commit the initial version of the target-cfg header
for nrf5340 platform. At the initial state, this
header only specifies the core platform-specific API,
not the API for peripheral security configuration.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

**_Edit: duplicated this header for nRF9160._**

ioannisg committed:
platform: nrf5340: target-cfg header

Commit the initial version of the target-cfg header
for nrf5340 platform. At the initial state, this
header only specifies the core platform-specific API,
not the API for peripheral security configuration.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
target-cfg header contains the platform-specific, internal CORE API to do security configuration on the TrustZone-enabled core. 

For (I suppose) historical reasons, this API is _still_ implemented for each of the Cortex-M33 platforms as a non-agnostic API, even though it contains generic functionality. I suppose the reason is that the same header contains security configuration for peripherals and _that_ one is almost always vendor-specific.

Therefore, I chose to follow the same path here, and introduce the nRF5340 internal target-cfg API in the same way. Note that the initial version does not introduce the API for peripheral configuration. We can do this at a later stage. This header now only bring the minimum that is needed.

anangl commented on platform/ext/target/nrf5340/target_cfg.h:
```diff
@@ -0,0 +1,95 @@
+#ifndef __TARGET_CFG_H__
+#define __TARGET_CFG_H__
+
+#include "platform/ext/common/uart_stdout.h"
```
Do we actually need this inclusion here?

ioannisg replied:
I guess so - it is common inclusion for target-cfg across different ports. I kept it for this reason.

anangl replied:
But maybe that inclusion was just blindly copied when new targets were added (and some old ones were used as templates)? Wouldn't it be better to add it when it appears necessary?

anangl commented on platform/ext/target/nrf5340/target_cfg.h:
```diff
@@ -0,0 +1,95 @@
+#define __TARGET_CFG_H__
+
+#include "platform/ext/common/uart_stdout.h"
+#include "tfm_plat_defs.h"
```
`<tfm_plat_defs.h>` would be more appropriate.

ioannisg replied:
I copied how it is done for other platform-specific headers

anangl replied:
`""` means: first look in the same directory, then in include paths, `<>` means start from include paths That's why I thought the latter would be more appropriate. But that's just a detail, I won't insist on it.

oyvindronningstad commented on platform/ext/target/nrf5340/target_cfg.h:
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2017-2019 Arm Limited
```
Update copyright/add our own?

ioannisg replied:
Not really update - maybe add own, yes..

ioannisg replied:
Updated :)

anangl commented on platform/ext/target/nrf5340/target_cfg.h:
```diff
@@ -0,0 +1,95 @@
+ *
+ * \return  Returns error code.
+ */
+int32_t spu_init_cfg(void);
```
I can't find that one declared/defined for other targets nor any references to it. Same story with `spu_clear_irq`. Are these indeed needed?

ioannisg replied:
I think so. they are defined, for example, for AN521 platform, for the MPC, PPC controllers.
I guess it is a way to enable the interrupt and initialize the security unit. Might well be empty functions for us. I kept them for now.

anangl replied:
Okay, I see my mistake. I was looking for that exact name, but this function should reflect the actual name of the controller, which is other for our target than for all the others. Sorry for the noise.

ioannisg replied:
that's fine - thanks for looking around the details

ioannisg commented:
@oyvindronningstad what worries me a bit is that this header should be duplicated for nrf9160 - i wonder if we should merge them to one...
RIght now each target dyplicates this target-cfg.h in the platform directory. Funny that all the MS2 targets are having almost identical copies of this header...

oyvindronningstad commented:
> @oyvindronningstad what worries me a bit is that this header should be duplicated for nrf9160 - i wonder if we should merge them to one...

Less copy-paste is better IMO :).

anangl commented on platform/ext/target/nrf5340/target_cfg.h:
```diff
@@ -0,0 +1,96 @@
+#include "platform/ext/common/uart_stdout.h"
+#include "tfm_plat_defs.h"
+
+
```
You should add also definitions of `TFM_DRIVER_STDIO` and `NS_DRIVER_STDIO`, see [readme.rst](https://github.com/NordicPlayground/trusted-firmware-m/tree/master/platform/ext#target_cfgh).

ioannisg replied:
Yeap, that's correct; I only wanted to do these later, once we have the names of these stdio drivers. My intention for now is to push whatever is periperhal-agnostic :)

ioannisg replied:
@anangl is it OK to not define these macros now, since we don't know how they will be called and do it after we have defined the respective driver struct instances?

anangl replied:
It is OK. We could also start with both those macros just set to `Driver_USART0`.

ioannisg replied:
OK  just brought back these defines, @anangl

ioannisg commented:
> > @oyvindronningstad what worries me a bit is that this header should be duplicated for nrf9160 - i wonder if we should merge them to one...
> 
> Less copy-paste is better IMO :).

I agree. For now, I committed the "framework", i.e there's a nordic_nrf parent directory, inside which there's the nrf5340 directory for the nRF53 target. For now, I want to keep the platform-cfg.h there, because I anticipate that it might not be identical for nrf5340 and nrf9160. If we see that they are identical, we can merge them and place them into a nordic_nrf/common/ folder

ioannisg committed:
platform: nrf9160: target-cfg header

Commit the initial version of the target-cfg header
for nrf9160 platform. At the initial state, this
header only specifies the core platform-specific API,
not the API for peripheral security configuration.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
